fvm/src/gas/price_list.rs

```diff
@@ -1074,56 +1110,277 @@ impl Rules for WasmGasPrices {
+                linear_cost($base, $linear, $multiplier)
+            };
+        }
+
+        macro_rules! charge_table {
+            ($($($op:ident),+$(,)? => $kind:ident ($($args:expr),*$(,)?),)*) => {
+                match instruction {
+                    $(
+                        $(| Operator::$op { .. })+ => {
+                            charge_inst!($kind($($args),*))
+                        },
+                    )*
+                }
             }
         }
 
         // Rules valid for nv16. We will need to be generic over Rules (massive
         // generics tax), use &dyn Rules (which breaks other things), or pass
         // in the network version, or rules version, to vary these prices going
         // forward.
-        match instruction {
-            // FIP-0032: nop, drop, block, loop, unreachable, return, else, end are priced 0.
-            Operator::Nop
-            | Operator::Drop
-            | Operator::Block { .. }
-            | Operator::Loop { .. }
-            | Operator::Unreachable
-            | Operator::Return
-            | Operator::Else
-            | Operator::End => Ok(InstructionCost::Fixed(0)),
-            // Memory related instructions...
-            Operator::TableInit { .. } | Operator::TableCopy { .. } => linear_cost(
-                self.exec_instruction_cost + self.memory_copy_base_cost,
+        charge_table! {
+            /******************/
+            /*  Control Flow  */
+            /******************/
+
+            // FIP-0032: nop, block, loop, unreachable, return, else, end are priced 0.
+            Nop, Block, Loop, Unreachable, Return, Else, End => free(),
+
+            Br       => fixed(self.jump_unconditional),
+            BrIf, If => fixed(self.jump_conditional),
+            BrTable  => fixed(self.jump_indirect + self.memory_access_cost),
+
+            // TODO M2.2: Charge to jump back, and charge for arguments.
+            Call          => fixed(self.jump_unconditional + self.call),
+            CallIndirect  => fixed(self.jump_indirect + self.memory_access_cost + self.call),
+
+            /**********************/
+            /*  Stack & Registers */
+            /**********************/
+
+            // Constants, stack ops, etc.
+            Drop,                                   // FIP-0032
+            I64ExtendI32U,                          // only unsigned is free.
+            I32WrapI64,                             // truncates
+            I32ReinterpretF32, I64ReinterpretF64,   // casts
+            F32ReinterpretI32, F64ReinterpretI64,   // casts other way
+            I32Const, I64Const, F32Const, F64Const, // inline constants
+            => free(),
+
+            // Locals (TODO M2.2). Free because these are register references.
+            LocalGet, LocalSet, LocalTee => free(),
+
+            // Globals (TODO M2.2). Free because these are register references.
+            GlobalGet, GlobalSet         => free(),
```

Comment:
The way this is written, wouldn't this apply the exemption to Skyr and Shark too?

Reply:
We probably need a `noop_equivalent` gas param that we can set to 4 in previous network versions.

Reply:
@Stebalien -- can you follow up here?

Reply:
Sorry, I never hit the submit button... See my reply above.

NOte: I'll merge this once I get a chance to merge the syscall gas changes.